### ISSUE TITLE
Use identifiable structs for AST variant fields

### ISSUE THREAD

**Issue by cmoog** [ballista-compute/sqlparser-rs#311]
When consuming the generated AST for further processing, one needs to match against and deconstruct `ast::Statement` instances. Currently, the fields of this enum's variants are defined inline... which makes it impossible for downstream code to represent the nested structures as types. For example, consider this structure:

https://github.com/ballista-compute/sqlparser-rs/blob/35ef0eee3857c59827b2587bc813218a00dce264/src/ast/mod.rs#L490-L497

I propose we migrate to the following pattern.
```rust
pub struct AnalyzeStatement {
  table_name: ObjectName, 
  partitions: Option<Vec<Expr>>, 
  for_columns: bool, 
  ....
}
pub enum Statement { 
  Analyze(AnalyzeStatement),
  ...
}
```

Obviously this is a massive breaking change, but that seems OK given the `0.9.x` crates.io version. 

Thoughts?

**Comment by alamb** [comment]
Is the primary usecase of this proposal so that downstream code can re-use `AnalyzeStatement` directly?

**Comment by cmoog** [comment]
> Is the primary usecase of this proposal so that downstream code can re-use AnalyzeStatement directly?

Yes, exactly.

**Comment by tvallotton** [comment]
I totally agree with this proposal. This would simplify my code considerably. I've got structs that are complete translations of some enum variants.

**Comment by tvallotton** [comment]
I forked the project and made those changes if you want to see how it would look.

**Comment by alamb** [comment]
I don't feel qualified to approve / accept this level of change without some additional buy in from @nickolay  / @andygrove  / @Dandandan . https://github.com/sqlparser-rs/sqlparser-rs/issues/311#issuecomment-902861958 seems like a good usecase to me but it is a pretty major change for all users of sqlparser-rs.

**Comment by nickolay** [comment]
https://github.com/sqlparser-rs/sqlparser-rs/issues/40 could be marked as a duplicate of this issue as well. As mentioned there, I didn't think creating a separate struct for _every single variant_ (even those holding one or two fields) of every enum was worth the breakage, but extracting complex variants like `Statement::Analyze` is something we have done before.

In https://github.com/sqlparser-rs/sqlparser-rs/pull/315#issuecomment-916948007 @Dandandan suggested doing this for all variants of the `Statement` enum, which seems a sensible suggestion too, as most statements exceed the complexity threshold anyway.

**Comment by alamb** [comment]
Your guidance makes a lot of sense to me @nickolay  -- thank you
